packages/yarnpkg-core/sources/Configuration.ts

```diff
@@ -1330,16 +1337,24 @@ export class Configuration {
     }
   }
 
-  useWithSource(source: string, data: {[key: string]: unknown}, folder: PortablePath, opts?: {strict?: boolean, overwrite?: boolean}) {
+  private getPackageEnv(packageCwd: PortablePath | null) {
+    if (!packageCwd) return {};
+    const envAbsolutePath = ppath.resolve(packageCwd, `.env` as PortablePath);
+    if (!xfs.existsSync(envAbsolutePath)) return {};
+    const envFile = xfs.readFileSync(envAbsolutePath, `utf8`);
+    return parseEnv(envFile);
+  }
+
+  useWithSource(source: string, data: {[key: string]: unknown}, folder: PortablePath, opts?: {strict?: boolean, overwrite?: boolean}, env?: DotenvParseOutput) {
```

Comment:
Can you move `env` inside `opts`?

Reply:
Should also use a different type, the fact that we're using `dotenv` should just be an internal implementation detail.

Reply:
@arcanis Of course~ https://github.com/yarnpkg/berry/pull/4835/commits/44c9b4da32383293391ab63318ed04e13716b3ba

@merceyz That's a good point! https://github.com/yarnpkg/berry/pull/4835/commits/6346bd7c7ef004ac00cb2550e490cb9e6ba984b9